### PR TITLE
fix(eslint-plugin): [lines-around-comment] extend rule options instead of overriding them

### LINKED ISSUE #6636: Bug: [lines-around-comment] afterHashbangComment option is not supported
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.1-rc&sourceType=module&code=MQQgZg9hBQBuCGAnABAE3gT2QXmQIgHNEBTeAFz2iA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RplWgQ2gHtYTACa0yfALbjELdGADa4bDkTR+0SABpFSrEuyRuAM3wqAEt2QALAEbcmAcwDCEqTIz44ibdgC+3gLqKfj5AA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
#!foo
var day = "great"
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/lines-around-comment": [
      "error",
      {
        "afterHashbangComment": true
      }
    ]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Should not crash, should warn about missing blank line after hashbang comment

### Actual Result

Configuration for rule "@typescript-eslint/lines-around-comment" is invalid:
        Value {"afterHashbangComment":true,"beforeBlockComment":true,"afterBlockComment":false,"beforeLineComment":false,"afterLineComment":false,"allowBlockStart":false,"allowBlockEnd":false} should NOT have additional properties.

### Additional Info

The playground example doesn't crash, but CLI crashes.

### PR DESCRIPTION
**Pull request by kozlovvski**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6636
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adjusts `lines-around-comment` schema to extend base rule properties instead of overriding them.

Additionally:
- creates `extendSchema` util, which uses `deepMerge` underneath and returns the merged schema in the same format as input -> either wrapped in array or not (however I am not sure why some rules wrap schema in braces and some don't). This could help avoid duplication and rule file bloating, like in `key-spacing`:
```ts
// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
const baseSchema = Array.isArray(baseRule.meta.schema)
  ? baseRule.meta.schema[0]
  : baseRule.meta.schema;
const schema = util.deepMerge(
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument -- https://github.com/microsoft/TypeScript/issues/17002
  baseSchema,
  {
    properties: {
      overrides: {
        properties: {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
          type: baseSchema.properties.overrides.properties.import,
        },
      },
    },
  },
);
```
shortened to:
```ts
const schema = util.extendSchema(baseRule.meta.schema,
  {
    properties: {
      overrides: {
        properties: {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
          type: baseSchema.properties.overrides.properties.import,
        },
      },
    },
  },
);
```
- adds `AST_TOKEN_TYPES.Shebang`. I've noticed, that we're missing this in typings. Adding this highlighted some places, where additional edge cases should be supported e.g. `ban-tslint-comment`

@bradzacher let me know if this is what you had in mind in https://github.com/typescript-eslint/typescript-eslint/issues/6636#issuecomment-1477505325

### CODE REVIEW

File: packages/eslint-plugin/src/rules/lines-around-comment.ts [hunk 1]
```diff
@@ -448,6 +394,8 @@ export default util.createRule<Options, MessageIds>({
                 before: options.beforeBlockComment,
               });
             }
+          } else if (token.type === AST_TOKEN_TYPES.Shebang) {
+            // do we want to do anything here?
           }
```
Comment (kozlovvski) [inline packages/eslint-plugin/src/rules/lines-around-comment.ts:399]: ESlint core checks for empty line here:
https://github.com/eslint/eslint/blob/main/lib/rules/lines-around-comment.js#L456

But I think we don't need to do anything here – shebang comments should trigger parser errors if used anywhere beside the first line and first line is covered by the base rule.

File: packages/eslint-plugin/typings/eslint-rules.d.ts [hunk 2]
```diff
@@ -759,6 +759,7 @@ declare module 'eslint/lib/rules/lines-around-comment' {
         allowModuleEnd?: boolean;
         ignorePattern?: string;
         applyDefaultIgnorePatterns?: boolean;
+        afterHashbangComment?: boolean;
```
Comment (kozlovvski) [inline packages/eslint-plugin/typings/eslint-rules.d.ts:762]: I am not sure if our additional options should belong here – when I first read through this I thought base ESlint rule supports options like `allowInterfaceStart`. IMO only base options should belong here and we should extend them in the rule file. WDYT?

**Comment by kozlovvski** [comment]
Obviously this is missing tests, but please tell mi if I am heading in the right direction 😄

File: packages/eslint-plugin/src/rules/lines-around-comment.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/lines-around-comment.ts:399]: Agreed, yeah, this looks good to me to leave out.

File: packages/eslint-plugin/src/rules/ban-tslint-comment.ts [hunk 3]
```diff
@@ -36,6 +36,11 @@ export default util.createRule({
       Program(): void {
         const comments = sourceCode.getAllComments();
         comments.forEach(c => {
+          // edge case to support
+          if (c.type === AST_TOKEN_TYPES.Shebang) {
+            return;
+          }
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/ban-tslint-comment.ts:42]: Could you add a unit test for this and any other rule that calls `sourceCode.getAllComments` please? Now that we support shebangs, any rule that interacts with all comments should be tested to make sure they don't cause it to crash, report erroneously, etc.

File: packages/eslint-plugin/typings/eslint-rules.d.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/typings/eslint-rules.d.ts:762]: Agreed, feels off for that reason to put them here.

**Review by JoshuaKGoldberg — COMMENTED** [review]
@bradzacher was talking about hashbang parsing support in #6500 so I will defer to him - just leaving some comments now to try to be useful in the meantime! ❤️

File: packages/eslint-plugin/src/rules/ban-tslint-comment.ts (hunk 3 above)
Comment (kozlovvski) [reply packages/eslint-plugin/src/rules/ban-tslint-comment.ts:42]: Yeah, I can do that 👍

**Comment by bradzacher** [comment]
So ESLint provides its own hashbang parsing. When ESLint sees file contents starting with `!#` it replaces it with `//` and then after parsing it updates the first comment in the `program.comments` array to have `type = 'Shebang'` (instead of `type = 'Line'`).

Kind-of annoying, but that was their hack back in the day to support hashbangs well before they were an official JS standard.

With that being said - we shouldn't be including `ShebangComment` in our AST spec - we don't produce this node ever, ESLint does indirectly. By adding it to the AST spec we're polluting our generic parser with ESLint-specifics - which we want to avoid!

It's going to be a larger and more complex change to make that work. In a nutshell we need to create the type within our ESLint types (located in `utils/ts-eslint`) and then update all the relevant spots to accept it in addition to the existing AST types.

If you're up for it happy to have you take a swing at this! Otherwise LMK if not - all good and i'll put up a PR that you could rebase on top of.

**Comment by kozlovvski** [comment]
Hey, I am closing this for now – I have trouble finding time recently and I don't want to keep a stale PR open.

**Comment by JoshuaKGoldberg** [comment]
No worries, thanks for letting us know! If you end up having time, great! If not / in the meantime, this can serve as a great reference for anybody else who's interested in tackling the issue.

For anybody else interested - please do go ahead. If you build off this code, we'll try to remember to ensure there's a [co author credit](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) on the PR.

**Pull request closed without merging**